### DISCUSSION TITLE
data_custom_control

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#38] [Announcements]
用户自定义控件

https://lv.qizhen.xyz/data_custom_control

**Comment by Arita9527** [comment]
老师，.net构造器为什么我导入DLL之后，没有看到方法和属性阿。

还有个问题，
1.怎么区别DLL是ACTIVEX还是.NET的呢？
2.没有头文件，LABVIEW可以用DLL吗？怎么知道每个函数的定义？

**Reply by ruanqizhen** [reply]
DLL 用于把一组函数打包成一个文件。ActiveX 和 .NET 是两种规范，它们定义了一个组件中需要包含哪些函数，每个函数有哪些参数，实现什么样的功能。根据 DLL 中都有哪些函数，就可以确定它是按照什么规范实现的了。
如果没有头文件，也可以通过查看文档，或是示例来了解 DLL 文件中每个函数的名称和参数定义。如果只是拿到一个别人内部的DLL，没有任何相关文档，是无法使用的。

**Comment by Arita9527** [comment]
老师，打开仪器的范例时
它提示让我去找.NET的DLL
1.如果我指定到仪器安装目录下的DLL，范例的方法和属性都会全部消失。如图
![](https://postimg.cc/p9TnhNkG)
为什么呢？我使用构造器加载DLL失败了？

**Comment by Arita9527** [comment]
老师，打开仪器的范例时
它提示让我去找.NET的DLL
1.如果我指定到仪器安装目录下的DLL，范例的方法和属性都会全部消失。如图
![构造器加载失败](https://i.postimg.cc/DzQQNnGT/PMengine-Erro.png)
为什么呢？我使用构造器加载DLL失败了？

**Reply by ruanqizhen** [reply]
这个问题我没法验证，所以也能太确定。看上去像是DLL加载失败，有可能依赖的一些底层库没找到。如果不用LabVIEW，有其它程序可以测试准格尔DLL吗？
把DLL放到系统目录下试试？

**Reply by Arita9527** [reply]
但是如果我不制定，让他自己找，反而不会报错。方法和属性都在。

系统目录，你指的是system32？

**Reply by Arita9527** [reply]
还有就是，我打包成EXE，运行EXE，会报错。
但是我把：
data删除（里面包括了仪器的DLL）
EXE放到仪器的安装目录下，
却可以运行。

**Reply by ruanqizhen** [reply]
system32 是一个系统目录，.NET也可能还有其它系统目录。LV自己寻找会在系统目录和LabVIEW的路径中寻找，它如果能找到合适的DLL，就可以正常运行。 打包成EXE，一般需要需要所有的DLL还在系统目录下，或者在EXE的目录下才能找到。

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：

## 在 LabVIEW 中使用常量定义

如下图所示，在C语言里，使用#define来定义一个常数是非常基本的用法。直接使用数字，时间一长，就不只到这个数字是哪来的了。而且，这种方法也便于修改在程序中多处使用的常量的值。在C++一般是用const来达到同样的目的。

![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFQyw8YsQz_CNdv2WP6yXEX9zoDBJbzcmRhsmTst48pOz-95pzruFSnuy5jQkM9zhk8bvHP-S4YFxoJmeM5SS6vnwf-oEXli00YWsOJH5rVHoaoo-4csU-5E)

**图1：C 语言中的常量定义**

    我以前在LabVIEW中编程，还从没注意过这个问题。一般哪里要用一个常数，直接就放一个constant在那里。如图2。

![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFQyw8YsQz_CNFaaHRniJVLtzYZGsVprpOX6pR1rIlu7d-3GT16EE6rbVODOXDsxJcUEb2jvr94t-soSBPjlfkc7CdT0vz5-zUTlTW5qahPu5zOPDdMep3yk)

**图2：在 LabVIEW 中使用常量的最普遍方法**

    以前编写的LabVIEW程序都比较小，一般是一个人开发的，所以这样写，也没有太大的麻烦。现在编写的程序规模越来越大，最近做的一个项目，VI数量已经上千了，有4个人参与编程。程序规模大了，不规范就很难维护。所以开始考虑这个问题。

    但是LabVIEW里面没有类似的功能，不知道为什么以前没人提意见？

    下面提出几种不算太完美，但有所进步的解决方案。

    一种简单的替代方法是使用type define control，自定义一个Ring control。关于 Type Def 的详细信息，可以参考《[用户自定义控件中 Control, Type Def. 和 Strict Type Def. 的区别](http://ruanqizhen.spaces.live.com/blog/cns!5852D4F797C53FB6!1825.entry)》。把要使用的常数作为Ring的值，给他个有意义的文字标签。在需要时用常数的地方，把这个带type define 的ring常数放上去，而不是直接放数值常量。这样就解决了上面提到的一个问题：可以有自带的文字说明。如图3所示。

![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFbYmmfzv-dncCZy1gXbcmbNW9QSKSJ1Sq8zEygAtOD0JlSo0opDMsjwezeTVB6-WywMP57xUBSdSkiG2KlrPHT6szaIInd_3tNDRYONPa06xaVIdeKjVozY)

**图3：利用 Type Def Ring 的解决方案**

    但是这样做还是有很多缺陷。首先是统一修改数值的问题。在自定义Ring中修改某一项的值，相关的常量不会跟着一起更新；还有一个缺陷是 Ring control 不支持多个标签是用同一数值；另外 Ring control 也没办法像 C 语言中一样使用表达式定义值。

    一个改进版的解决方案是使用 Enum Type Def 把所有常量名字列出来，再写一个 VI 用于得到常量的真实值，如图4所示。这样解决了不同标签可以返回相同值的问题，也可以自动更新常量值，但是使用表达式还是不方便。

![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFbYmmfzv-dncIld2ChfSA2uYsHQeltwjqFLtm-HrrkBH8NGwB9zHSLwpwON94dsYz9y09qll5pqVzYzRL0QR0OkxW0n5Sx9hvGbRB5S7RvrxclX6PQsujfY)

**图4：利用 Enum Type Def 和 subVI**

    我目前在程序中使用的方法是，把所有要用到的常量，全部做成全局变量。全局变量可以用 Global ，但我喜欢用 VI 全局变量。就是把变量记载 shift regisiter 中。然后，用一个初始化的 VI 负责在程序运行开始时初始化所有的全局变量。这样，以后如果需要更改某一常数值，就只需改这一个VI就可以了。

    不过，现在回想，还是用 Global 好一些。我以前测试过，Global 读写的速度比 VI 要慢很多，所以我不喜欢 Global 。但是，常量值在程序中用的并不频繁，所以速度不是个问题。但是数量很多，用 VI 表示就不太合适了，每个常数都要创建一个 VI 非常费事。另一个缺点是如果在后面板换用一个常量，还要再拖另一个 VI 上来，很麻烦。用 Global 会好一些，但还不是让我太满意。

    要想有一个完美的解决办法，只能再造一个新东西了。@#$%^&\* （此处属公司机密，删去256字）

**Reply by ruanqizhen** [reply]
之前博客下的留言：

mcusir
2011年05月14日 00:07 编辑
我来报下料，NI公司的机密都是将别人家的技巧据为己有，常量使用WORM-Global,建议读下这个网址：
http://forums.ni.com/t5/LabVIEW/WORM-Global/m-p/1545910

回复
Qizhen
2006年04月9日 07:41 编辑
一个常量子VI中设置多个常量也不是好办法，一个VI最多设置六个输出。再多，连线就乱七八糟了，我是很重视程序框图干净整洁的。

回复
竹贤
2006年04月9日 07:34 编辑
可以在一个常量子VI中设置多个常量。

回复
yoyo
2006年04月4日 19:08 编辑
舅舅机滚的挖，乖乖！

回复
冰雪儿
2006年04月2日 17:22 编辑
这里真得有好多要学的东西啊,我会常来的! ^ – ^ !!!

回复
竹贤
2006年04月1日 08:52 编辑
也可以把多个常量作为一个子程序，在需要常量的地方放入这个输出常量的子程序即可。要修改常量只需修改这个子程序即可。

回复
jing
2006年03月31日 14:33 编辑
我又来凑下热闹先~~~呵呵~~计算机系我既死穴~~~

回复

**Reply by ruanqizhen** [reply]
之前博客下的留言：

陈小刚
2011年03月30日 16:26 编辑
用严格类型定义的时候，似乎有一些控件其实例的常量也是与定义控件相关的，比如枚举型控件

回复
ruanqizhen
2011年03月31日 09:23 编辑
确实是这样的，现在的LabVIEW在这上面有点乱。Enumerate常量会变，而Ring常量不会变。

回复
